### PR TITLE
feat: add middleware destination

### LINKED ISSUE #521: Root _middleware is called for every request
I'm working on a session middleware which is called in the `routes/_middleware.ts` to be able to access the session across the app.

As I see it, the root `_middleware` is called for every request even the one outside the routes (e.g. static resource request, _frsh/alive, etc.) which could be the intended behavior as it's consider a global middleware.

My issue is that with this behavior, it's impossible to keep a session initialization in the root middleware and implement features like session flash as the multiple request expires the flashed data.

It could be possible to filter out from the request, but seems pretty unreliable.

Is it the expected behavior ? If so, maybe `routes/_middleware` could be reserved only for request routed to a route, and we could find another way to implement global middlewares ?

### PR DESCRIPTION
**Pull request by crowlKats**
Closes #521

### CODE REVIEW

File: src/server/router.ts [hunk 1]
```diff
@@ -30,16 +38,22 @@ export interface Routes<T = {}> {
   [key: string]: { [K in KnownMethod | "default"]?: MatchHandler<T> };
 }
 
+export type RouteKinds = "internal" | "static" | "route" | "notFound" | "error";
```
Comment (lucacasonato) [inline src/server/router.ts]:
```suggestion
export type RouteKind = "internal" | "static" | "route" | "notFound" | "error";
```

**Review by lucacasonato — COMMENTED** [review]
Rather than `routeKind` can you make it `destination` (this aligns with fetch destination in browsers).

**Review by lucacasonato — COMMENTED** [review]
Direction is much better now. Thanks.

I have some suggestions for simplification:

instead of doing:

1. call inner with (req, ctx, handlers, middlewareCtx), perform routing, getting the handler
2. have inner add the handler to `handlers`
3. call `middlewareCtx.next()` in `inner` 

do instead:

1. call `inner` with (req, ctx), perform routing, getting the handler
2. return from `inner` the `handler` and the `targetKind`
3. add the handler to handlers _outside_ of inner
4. add the target to middlewareCtx _outside_ of inner
5. call `middlewareCtx.next()` _outside_ of inner

here `inner` function returned from `router` would return a `{ target: TargetKind, handler: () => {} }` on call. Inner can also be synchronous then.

This is important, because right now the error handling is very strange. There should not be an `"error"` targetKind, because when you call all the middlewares, you don't know if something is going to error yet. The act of routing itself should not be able to result in an error, only ever the execution (the handlers).

File: tests/main_test.ts [hunk 2]
```diff
@@ -494,6 +494,29 @@ Deno.test({
   },
 });
 
+Deno.test("middleware destination", async (t) => {
```
Comment (crowlKats) [inline tests/main_test.ts:497]: an `internal` test is missing, but unsure if there is any path that is consistent for an internal asset

Comment (lucacasonato) [reply tests/main_test.ts:497]: _frsh/refresh.js

Comment (crowlKats) [reply tests/main_test.ts:497]: done

**Review by lucacasonato — APPROVED** [review]
Excellent, thank you!

**Pull request merged**